File: CPP/src/Platypus.cpp
#include <iostream>
#include "../include/Platypus.hpp"

using namespace std;

/*
Constructor yang menerima parameter position, name, meat, dan milk serta egg
Parameter diturunkan dari MeatProducing dan MilkProducing dan EggProducing
*/
Platypus::Platypus(pair<int,int> _position, string _name, Land* _landPos){
    initObject(PLATYPUS, 'Y');
    initializeFarmAnimal(_position, _name, _landPos);
    initializeMeatProducing(new PlatypusMeat(HARGA_DAGING_PLATYPUS, getName()));
    initializeEggProducing(NULL);
    initializeMilkProducing(NULL);
    setIsProductProduced(true);
}

/*
Method untuk memproduksi daging Platypus. Method diturunkan dari
MeatProducing
*/
void Platypus::produceMeat(Player& _player){
    if (meat != NULL){
        _player.addInventory(new PlatypusMeat(HARGA_DAGING_PLATYPUS, getName()));
        delete meat;
    }
}

/*
Method untuk memproduksi susu Platypus. Method diturunkan dari
MilkProducing
*/
void Platypus::produceMilk(Player& _player){
    if (milk != NULL){
        _player.addInventory(new PlatypusMilk(HARGA_SUSU_PLATYPUS, getName()));
        delete milk;
    }
}

/*
Method untuk memproduksi egg Platypus. Method diturunkan dari
EggProducing
*/
void Platypus::produceEgg(Player& _player){
    if (egg != NULL){
        _player.addInventory(new PlatypusEgg(HARGA_TELUR_PLATYPUS, getName()));
        delete egg;
    }
}

void Platypus::eat(){
    //cout <<"testPlatypusMakan1"  << endl;
    FarmAnimal::eat();
    setIsProductProduced(false);
    //cout << "testPlatypusMakan" << endl;
    initializeEggProducing(new PlatypusEgg(HARGA_TELUR_PLATYPUS, getName()));
    initializeMilkProducing(new PlatypusMilk(HARGA_SUSU_PLATYPUS, getName()));
}

void Platypus::talk(){
    cout << name << ": Eyy, what the hell?" << endl;
}

void Platypus::interact(Player* _p){
    //cout << "PLATYPUS INTERACT" << endl;
    if (!getIsProductProduced()){
       // cout << "PLATYPUS INTERACT KEDUA" << endl;
        produceEgg(*_p);
        produceMilk(*_p);
        //cout << "PLATYPUS INTERACT KETIGA" << endl;
        setIsProductProduced(true);
    }
}

File: Java/src/main/java/engisubs/gameobject/farmanimal/Platypus.java
package engisubs.gameobject.farmanimal;

import java.util.*;

import engisubs.gameobject.Player;
import engisubs.gameobject.cell.land.*;
import engisubs.gameobject.product.farmproduct.meat.*;
import engisubs.gameobject.product.farmproduct.egg.*;
import engisubs.gameobject.product.farmproduct.milk.*;

public class Platypus extends FarmAnimal implements MeatProducing, EggProducing, MilkProducing{
    /**
     * Constructor dari Platypus
     * @param  _position Posisi dari Platypus
     * @param  _name     Nama dari Platypus
     * @param  _landPos  Posisi Land dimana Platypus berada
     * @return           Instance dari Platypus
     */    
    public Platypus(Map <String, Integer> _position, String _name, Land _landPos){
        super.killable = true;
        initGameObject(GameObjectType.PLATYPUS, 'Y');
        initializeFarmAnimal(_position, _name, _landPos);
    }
    /**
     * Method untuk menghasilkan daging
     * @param _p Reference dari player untuk melakukan penambahan inventory
     */
    public void produceMeat(Player<?> _p){
        _p.addInventory(new PlatypusMeat(70, name));
    }
    /**
     * Method untuk menghasilkan telur
     * @param _p Reference dari player untuk melakukan penambahan inventory
     */
    public void produceEgg(Player<?> _p){
        _p.addInventory(new PlatypusEgg(10, name));
    }
    /**
     * Method untuk menghasilkan susu
     * @param _p Reference dari player untuk melakukan penambahan inventory
     */
    public void produceMilk(Player<?> _p){
        _p.addInventory(new PlatypusMilk(15, name));
    }
    /**
     * Method untuk berbicara
     */
    public void talk(){
        System.out.println(name + ": It is I, " + name + ", Destroyer of Worlds!");
    }
    /**
     * Method untuk melakukan interaksi dimana Platypus akan menghasilkan telur dan susu
     * @param _p Reference dari player untuk melakukan penambahan inventory
     */
    public void interact(Player<?> _p){
        if (!getIsProductProduced()){
            produceEgg(_p);
            produceMilk(_p);
            isProductProduced = true;
        }
    }
    /*
    public void eat() {
        System.out.println("PLATYPUS MAKAN");
        super.eat();
        setIsProductProduced(false);
    }
    */
}